### DISCUSSION TITLE
IP Adapter for composition

### DISCUSSION THREAD

**Discussion by asomoza** [huggingface/diffusers#7419] [Show and tell]
Hi, there's a new IP Adapter that was trained by @jaretburkett to just grab the composition of the image.

Here's the release [tweet for SD 1.5](https://twitter.com/ostrisai/status/1769119254610096603) and for [SDXL](https://twitter.com/ostrisai/status/1770273049536807142) 

I think it works good when the model you're using understand the concepts of the source image. Also the scale and the CFG play an important role in the quality of the generation.

Here's my tests with it:

|source|a cat plushie|a cat|plushie in the desert|
|---|---|---|---|
|![20240215062951_1186234956](https://github.com/huggingface/diffusers/assets/5442875/5e029c3c-adc3-40c6-8da9-3918efa6cfd5)|![20240321074439_1964483001](https://github.com/huggingface/diffusers/assets/5442875/78353a49-31b2-47d3-8866-fb48d37c9014)|![20240321074156_1829664615](https://github.com/huggingface/diffusers/assets/5442875/48e11555-d198-40c1-a2bd-9be35c409269)|![20240321074707_1291586369](https://github.com/huggingface/diffusers/assets/5442875/55ee61bd-dbef-465e-8265-11aece1aed95)|

|source|a man|anthropomorphic bear|anime man wearing spacesuit|
|---|---|---|---|
|![woman_source](https://github.com/huggingface/diffusers/assets/5442875/e9839980-bca8-4b33-8402-486b1c49324a)|![20240321074830_1656593121](https://github.com/huggingface/diffusers/assets/5442875/f4568140-93b0-482c-ab4b-11aa4189eabc)|![20240321075158_1626876769](https://github.com/huggingface/diffusers/assets/5442875/7b0ef435-24ad-4477-b5bf-11af0a094a98)|![20240321075306_1905411366](https://github.com/huggingface/diffusers/assets/5442875/7db664aa-8c1e-4743-ab47-96a768de8258)|

Here's the code I used if someone wants to test it too, even better if you can share your results and findings here too.

```py
import torch
from transformers import CLIPVisionModelWithProjection

from diffusers import DPMSolverMultistepScheduler, StableDiffusionXLPipeline
from diffusers.utils import load_image


image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", subfolder="models/image_encoder", torch_dtype=torch.float16
).to("cuda")

pipeline = StableDiffusionXLPipeline.from_pretrained(
    "RunDiffusion/Juggernaut-XL-v9", torch_dtype=torch.float16, image_encoder=image_encoder, variant="fp16"
).to("cuda")

pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)
pipeline.scheduler.config.use_karras_sigmas = True

pipeline.load_ip_adapter(
    "ostris/ip-composition-adapter",
    subfolder="",
    weight_name=[
        "ip_plus_composition_sdxl.safetensors",
    ],
    image_encoder_folder=None,
)
pipeline.set_ip_adapter_scale([0.8])

ip_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/20240215062951_1186234956.png?download=true"
)

prompt = "a cat plushie"

image = pipeline(
    prompt=prompt,
    negative_prompt="",
    ip_adapter_image=[ip_image],
    guidance_scale=6.5,
    num_inference_steps=25,
).images[0]

image.save("result.png")
```

Pictures are either generated by me or grabbed from https://www.publicdomainpictures.net/

**Comment by asomoza** [comment]
I've done more experiments with this adapter:

## Composition with styles and masks

If we have an image of a couple for composition and we want to apply a style for each person, we can do it with 3 IP Adapters:

|Composition image|male style|female style|Final image|
|---|---|---|---|
|![young-couple-isolated-14868229109Pl](https://github.com/huggingface/diffusers/assets/5442875/0114c1ed-21bc-460a-9af3-65c959921ede)|![vintage-santa-claus-postcard-1698683643O1W](https://github.com/huggingface/diffusers/assets/5442875/ef32f576-2f28-4272-872e-5f8824ec4fb6)|![woman](https://github.com/huggingface/diffusers/assets/5442875/df602eec-a6e5-4473-a539-04d7eb5e9ffd)|![20240321125857](https://github.com/huggingface/diffusers/assets/5442875/f29049e1-5a08-4ffc-b7ee-f91bf7b41a7a)|

The final image I get is without any prompt. If I add a prompt, another ip adapter for the background and play with the scale and noise of each one, I got this final image:

![20240321131043](https://github.com/huggingface/diffusers/assets/5442875/e538d8cb-4907-4903-9b0b-904e61097343)

**Reply by yiyixuxu** [reply]
wow! 
did you have to use a mask here? or the composition adapter is enough? can you share the code?

**Reply by asomoza** [reply]
I used a mask for her, him and the background, I used my app because it is faster to do experiments with it. I'll do it with just code and post it here.

**Reply by asomoza** [reply]
I had to use a white image to do a fake mask for the composition adapter, because the IPAdapterMaskProcessor doesn't accept a None in the list.

Finally I can't seem to get the photorealistic result without the negative noise, it always keep the cartoonish style:

|||
|---|---|
|![20240321150910_1399586295](https://github.com/huggingface/diffusers/assets/5442875/bdff97bf-0bc9-45ed-9a3e-488cec74e75f)|![result](https://github.com/huggingface/diffusers/assets/5442875/9a31d988-f193-4704-815d-06413b3df51b)|

To get the photorealistic results is necessary to do what I did  in this discussion: https://github.com/huggingface/diffusers/discussions/7167.

Anyways, here's the code:

```py
import torch
from transformers import CLIPVisionModelWithProjection

from diffusers import DPMSolverMultistepScheduler, StableDiffusionXLPipeline
from diffusers.image_processor import IPAdapterMaskProcessor
from diffusers.utils import load_image, logging
from diffusers.utils.logging import set_verbosity


set_verbosity(logging.ERROR)  # to not show cross_attention_kwargs...by AttnProcessor2_0 warnings

image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", subfolder="models/image_encoder", torch_dtype=torch.float16
).to("cuda")

pipeline = StableDiffusionXLPipeline.from_pretrained(
    "RunDiffusion/Juggernaut-XL-v9", torch_dtype=torch.float16, image_encoder=image_encoder, variant="fp16"
).to("cuda")

pipeline.scheduler = DPMSolverMultistepScheduler.from_config(pipeline.scheduler.config)
pipeline.scheduler.config.use_karras_sigmas = True

pipeline.load_ip_adapter(
    ["ostris/ip-composition-adapter", "h94/IP-Adapter", "h94/IP-Adapter", "h94/IP-Adapter"],
    subfolder=["", "sdxl_models", "sdxl_models", "sdxl_models"],
    weight_name=[
        "ip_plus_composition_sdxl.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
    ],
    image_encoder_folder=None,
)
pipeline.set_ip_adapter_scale([1.0, 0.75, 0.75, 0.3])

ip_composition_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter__20240321125152.png"
)
ip_female_style = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter__20240321125625.png"
)
ip_male_style = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter__20240321125329.png"
)
ip_background = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter__20240321130643.png"
)

composition_mask = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/1024_whole_mask.png"
)
female_mask = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter_None_20240321125641_mask.png"
)
male_mask = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter_None_20240321125344_mask.png"
)
background_mask = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/ip_adapter_6_20240321130722_mask.png"
)


prompt = "high quality, cinematic photo, cinemascope, 35mm, film grain, highly detailed"
negative_prompt = "anime, cartoon"


processor = IPAdapterMaskProcessor()
masks = processor.preprocess([composition_mask, female_mask, male_mask, background_mask], height=1024, width=1024)

image = pipeline(
    prompt=prompt,
    negative_prompt="",
    ip_adapter_image=[ip_composition_image, ip_female_style, ip_male_style, ip_background],
    cross_attention_kwargs={"ip_adapter_masks": masks},
    guidance_scale=6.5,
    num_inference_steps=25,
).images[0]

image.save("result.png")

```

**Reply by yiyixuxu** [reply]
ohh you won't have to copy the weights over, you can do this 
```python
pipeline.load_ip_adapter(
    ["ostris/ip-composition-adapter", "h94/IP-Adapter", "h94/IP-Adapter", "h94/IP-Adapter"],
    subfolder=["", "sdxl_models", "sdxl_models", "sdxl_models"],
    weight_name=[
        "ip_plus_composition_sdxl.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
    ],
)
```

**Reply by asomoza** [reply]
oh didn't know that, I'll fix it.

**Reply by yiyixuxu** [reply]
also, it's a little bit inefficient to load the same ip-adapter 3 times, no?
is it possible to do something like this? if it is not possible with our current implementation, what needs to be done to support this? 
```python
pipeline.load_ip_adapter(
    ["ostris/ip-composition-adapter", "h94/IP-Adapter"],
    subfolder=["", "sdxl_models",],
    weight_name=[
        "ip_plus_composition_sdxl.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
    ],
)

pipeline.set_ip_adapter_scale([1.0, [0.75, 0.75, 0.3]])
...
image = pipeline(
    prompt=prompt,
    negative_prompt="",
    ip_adapter_image=[ip_composition_image, [ip_female_style, [ip_male_style, ip_background]],
    cross_attention_kwargs={"ip_adapter_masks": masks},
    guidance_scale=6.5,
    num_inference_steps=25,
).images[0]


```

**Reply by asomoza** [reply]
I'll test it, this was just a quick conversion from my code, so if you find any other mistakes, I'll fix them

**Reply by asomoza** [reply]
I tested it with `pipeline.set_ip_adapter_scale([1.0, [0.75, 0.75, 0.3]])` and it doesn't work and I got this error:  ` ValueError: ip_adapter_image must have same length as the number of IP Adapters. Got 4 images and 2 IP Adapters.` The set_ip_adapter_scale doesn't do anything with the second list, just sets the `attn_processor.scale`.

I can probably assign multiple images to the same adapter but I lose the ability to set different scales to each one since diffusers didn't implement the set different scales to each image and also with that I'll still have the problem that I can't use different mask with each image.

There're several ways for it to work but I don't know which one is better for diffusers. This is the reason that made me code my own node system instead of the pipelines, AFAIK this happens too if you want to use multiple controlnets/t2i adapters that are of the same type but with different images, I always found inefficient to load multiple times the same weights.

In my system I load the `ip adapter` once and afterwards I can use it with all the images and masks I want, same as ComfyUI.

I can do something to make it prettier like what was done in this comment: https://github.com/huggingface/diffusers/pull/6847#issuecomment-1932947474 but that doesn't solve the problem.

```py
pipeline.load_ip_adapter(
    ["ostris/ip-composition-adapter", "h94/IP-Adapter"],
    subfolder=["", "sdxl_models",],
    weight_name=["ip_plus_composition_sdxl.safetensors"] + ["ip-adapter_sdxl_vit-h.safetensors"] * 3,
)
```

I don't think it's going to be easy to make this work with the current implementation without doing a major refactor but maybe @fabiorigano could have a solution.

**Reply by fabiorigano** [reply]
hi @asomoza thanks for adding me 

I made some changes to support multiple masks with one IP Adapter. 

```py
pipeline.load_ip_adapter(
    ["ostris/ip-composition-adapter", "h94/IP-Adapter"],
    subfolder=["", "sdxl_models"],
    weight_name=[
        "ip_plus_composition_sdxl.safetensors",
        "ip-adapter_sdxl_vit-h.safetensors",
    ],
    image_encoder_folder=None,
)
```

First, I created the masks as follows (later I had to suppress the "ip_adapter_mask should be a tensor ..." ValueError):

```py
masks1 = processor.preprocess([composition_mask], height=1024, width=1024)
masks2 = processor.preprocess([female_mask, male_mask, background_mask], height=1024, width=1024)
masks2 = masks2.reshape(1, masks2.shape[0], masks2.shape[2], masks2.shape[3])  # output -> (1, 3, 1024, 1024)
masks = [masks1, masks2]
```

I set ip adapter scales (one for each image) and called the pipeline:
```
pipeline.set_ip_adapter_scale([1.0, [0.75, 0.75, 0.3]])
image = pipeline(
    prompt=prompt,
    negative_prompt="",
    ip_adapter_image=[ip_composition_image, [ip_female_style, ip_male_style, ip_background]],
    cross_attention_kwargs={"ip_adapter_masks": masks},
    guidance_scale=6.5,
    num_inference_steps=25,
).images[0]
```

Small change to IPAdapterMaskProcessor.downsample:

  ```py
      if mask.ndim == 3:
            mask = mask.unsqueeze(0)
        o_h = mask.shape[2]
        o_w = mask.shape[3]
...
        mask_downsample = F.interpolate(mask, size=(mask_h, mask_w), mode="bicubic").squeeze(0)
```

Finally, the changes to the AttentionProcessor class:
```py
        # for ip-adapter
        for current_ip_hidden_states, scale, to_k_ip, to_v_ip, mask in zip(
            ip_hidden_states, self.scale, self.to_k_ip, self.to_v_ip, ip_adapter_masks
        ):
            
            if mask is not None:
                if not isinstance(scale, list):
                    scale = [scale]
                for i in range(mask.shape[1]):
                    ip_key = to_k_ip(current_ip_hidden_states[:, i, :, :])
                    ip_value = to_v_ip(current_ip_hidden_states[:, i, :, :])

                    ip_key = ip_key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
                    ip_value = ip_value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)

                    # the output of sdp = (batch, num_heads, seq_len, head_dim)
                    # TODO: add support for attn.scale when we move to Torch 2.1
                    _current_ip_hidden_states = F.scaled_dot_product_attention(
                        query, ip_key, ip_value, attn_mask=None, dropout_p=0.0, is_causal=False
                    )

                    _current_ip_hidden_states = _current_ip_hidden_states.transpose(1, 2).reshape(
                        batch_size, -1, attn.heads * head_dim
                    )
                    _current_ip_hidden_states = _current_ip_hidden_states.to(query.dtype)

                    mask_downsample = IPAdapterMaskProcessor.downsample(
                        mask[:, i, :, :], batch_size, _current_ip_hidden_states.shape[1], _current_ip_hidden_states.shape[2]
                    )

                    mask_downsample = mask_downsample.to(dtype=query.dtype, device=query.device)
                    hidden_states = hidden_states + scale[i] * (_current_ip_hidden_states * mask_downsample)
            else:
                ip_key = to_k_ip(current_ip_hidden_states)
                ip_value = to_v_ip(current_ip_hidden_states)

                ip_key = ip_key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
                ip_value = ip_value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)

                # the output of sdp = (batch, num_heads, seq_len, head_dim)
                # TODO: add support for attn.scale when we move to Torch 2.1
                current_ip_hidden_states = F.scaled_dot_product_attention(
                    query, ip_key, ip_value, attn_mask=None, dropout_p=0.0, is_causal=False
                )

                current_ip_hidden_states = current_ip_hidden_states.transpose(1, 2).reshape(
                    batch_size, -1, attn.heads * head_dim
                )
                current_ip_hidden_states = current_ip_hidden_states.to(query.dtype)

                hidden_states = hidden_states + scale * current_ip_hidden_states
```

**Reply by fabiorigano** [reply]
I updated the AttentionProcessor class in my solution in the previous comment, now it works correctly.

![result](https://github.com/huggingface/diffusers/assets/57982783/2ecfd9bc-def8-4bd7-926b-330804f8d6bf)

it obviously runs much faster than using 3 copies of the same IP Adapter

**Reply by asomoza** [reply]
thank you @fabiorigano I think this could be a good solution, do you have it a as a branch in your fork so I can test it?

**Reply by fabiorigano** [reply]
just pushed a new branch [multiimageipmask](https://github.com/fabiorigano/diffusers/tree/multiimageipmask), let me know if it works properly for you

the update only affects IPAdapterAttnProcessor2_0

**Reply by asomoza** [reply]
Hi tested it and it works as intended, nice!. @yiyixuxu is this a solution we can add to diffusers? If so, I can then update this example afterwards.

**Reply by yiyixuxu** [reply]
@fabiorigano feel free to open a PR! IMO it is getting a little bit hard to read now but we can think about how to refactor from there
